src/net/server/middleware/processors/mandatory.rs

```diff
@@ -0,0 +1,340 @@
+        }) = request.transport()
+        {
+            let max_response_size_hint = *max_response_size_hint as usize;
+            let response_len = response.as_slice().len();
+
+            if response_len > max_response_size_hint {
+                // Truncate per RFC 1035 section 6.2 and RFC 2181 sections 5.1
+                // and 9:
+                //
+                // https://datatracker.ietf.org/doc/html/rfc1035#section-6.2
+                //   "When a response is so long that truncation is required,
+                //    the truncation should start at the end of the response
+                //    and work forward in the datagram.  Thus if there is any
+                //    data for the authority section, the answer section is
+                //    guaranteed to be unique."
+                //
+                // https://datatracker.ietf.org/doc/html/rfc2181#section-5.1
+                //   "A query for a specific (or non-specific) label, class,
+                //    and type, will always return all records in the
+                //    associated RRSet - whether that be one or more RRs.  The
+                //    response must be marked as "truncated" if the entire
+                //    RRSet will not fit in the response."
+                //
+                // https://datatracker.ietf.org/doc/html/rfc2181#section-9
+                //   "Where TC is set, the partial RRSet that would not
+                //    completely fit may be left in the response.  When a DNS
+                //    client receives a reply with TC set, it should ignore
+                //    that response, and query again, using a mechanism, such
+                //    as a TCP connection, that will permit larger replies."
+                //
+                // https://datatracker.ietf.org/doc/html/rfc6891#section-7
+                //   "The minimal response MUST be the DNS header, question
+                //     section, and an OPT record.  This MUST also occur when
+                //     a truncated response (using the DNS header's TC bit) is
+                //     returned."
+
+                // Tell the client that we are truncating the response.
+                response.header_mut().set_tc(true);
+
+                // Remember the original length.
+                let old_len = response.as_slice().len();
+
+                // Copy the header, question and opt record from the
+                // additional section, but leave the answer and authority
+                // sections empty.
+                let source = response.as_message();
+                let mut target = mk_builder_for_target();
+
+                *target.header_mut() = source.header();
+
+                let mut target = target.question();
+                for rr in source.question() {
+                    target.push(rr?)?;
+                }
+
+                let mut target = target.additional();
+                if let Some(opt) = source.opt() {
```

Comment:
This may fail if opt is very big.

Reply:
Yes. Pushing into the target can fail in many places in domain. In this case it would result in an error which would be caught at the call site:

```rust
        if let Err(err) = Self::truncate(request, response) {
            error!("Error while truncating response: {err}");
            *response = self.error_response(request, Rcode::ServFail);
            return;
        }
```

Do you have a suggestion for handling this differently?

Reply:
That is wrong. The result needs to be a truncate message, not ServFail. If the opt record doesn't fit, leave it out.

Reply:
But in this case it tried to truncate and failed wit an internal error. That is a SERVFAIL, or else how do you handle this?

Reply:
Well, the truncate is broken. It should truncate, and not fail while adding a big OPT record.

Reply:
I can log the push error and continue, but then RFC 6891 says:

> The minimal response MUST be the DNS header, question
> section, and an OPT record.  This MUST also occur when
> a truncated response (using the DNS header's TC bit) is
> returned.

So I have changed it to also attempt to add a minimal OPT record if pushing the actual OPT record fails.

Please see if that is better.

Reply:
Yes, I think that will work.